### PR TITLE
UX: Ask before editing a file

### PR DESCRIPTION
**Pull request by paroque28**
CLI will now ask you before editing a file and also what editor to use.
Example:
```bash
$ ./scripts/sh-services edit all

Applying command 'edit' to all services
1) amqp-stats  3) mcp         5) sio
2) live555     4) rabbitmq    6) Exit
Please enter the service you want to edit: 3
Will edit: mcp
###############################################
### mcp
###############################################
1) View service configuration  3) Exit
2) Edit service
Please enter your choice: 2
Editing /Users/pablorod/code/services/mcp
1) vim
2) nano
3) vi
4) emacs
Select your preferred text editor: 1
Do you want to edit the main configuration: /Users/pablorod/code/services/mcp/conf/0001-edit.env? (y/n)  
>  After merging with the default values, the outcome will be stored at <service>/.env.
> y
Merging /Users/pablorod/code/services/mcp/.env /Users/pablorod/code/services/mcp/conf/0001-edit.env /Users/pablorod/code/services/mcp/conf/custom.env /Users/pablorod/code/services/mcp/conf/default.env /Users/pablorod/code/services/conf/0010-arch.env into '/Users/pablorod/code/services/mcp/.env'

Resulting '/Users/pablorod/code/services/mcp/.env' file:

DB_PATH=/data/sighthound/db
LOG_PATH=/data/sighthound/logs/mcp
MCP_CONTAINER_NAME=mcp
MCP_DOCKER_IMAGE=us-central1-docker.pkg.dev/ext-edge-analytics/docker/mcp
MCP_DOCKER_TAG=1.2.4
MEDIA_PATH=/data/sighthound/media
SERVICE_GROUP=-dev
SH_ARCH=-amd64
SH_NODE=false
SIO_LICENSE_PATH=/data/sighthound/license/sighthound-license.json

Do you want to edit /Users/pablorod/code/services/mcp/conf/mcp.yml? (y/n) n
####################DONE#######################
1) amqp-stats
2) live555
3) mcp
4) rabbitmq
5) sio
6) Exit
Please enter your choice: 6
Exiting...
```

### CODE REVIEW

**Comment by w3sip** [comment]
I'm still wondering if our focus is in the right place. All of the above is useful for our internal consumption -- especially, when deploying on device and integrating with device's UI.

But for an end user, say Cox, or MPS on their server -- it is primarily a deployment template. I expect them to take it, learn from it -- and either come up with their own config set, or use as is, without ever upgrading services (but maybe upgrading the images or configuration).

For such a user, a complex infrastructure of scripts, choice of editors, etc will be an extra layer of complexity. For example, why would user want to use `deploy-services.sh`, rather than saying "hey, download this zip file from this URL, and deploy its contents in the folder of your choice"? What does the script buy for someone like Marcus at MPS in terms of usability?

**Comment by paroque28** [comment]
@w3sip sorry this PR is confusing, `deploy-services` is just the unmerged commit from https://github.com/sighthoundinc/services/pull/41
It has nothing to do with the sh-services or end-user UI. 

`deploy-services` is for our manufacturing team or a very very special request from a customer

**Comment by w3sip** [comment]
Should we have a separate private repo for it, then (perhaps, with `services` as a submodule)? I'd like to not give customers more content than they need to solve their problems? ;)

**Comment by paroque28** [comment]
Sure, I'll do that!

**Comment by paroque28** [comment]
I have removed the commit!

**Comment by paroque28** [comment]
@w3sip, this is ready to merge. The latest changes include documentation improvements

File: amqp-stats/conf/default.env [hunk 1]
```diff
@@ -2,6 +2,4 @@ AMQP_STATS_DOCKER_IMAGE=us-central1-docker.pkg.dev/ext-edge-analytics/docker/amq
 AMQP_STATS_DOCKER_TAG=1.2.3
 AMQP_STATS_CONTAINER_NAME=amqp-stats
 SERVICE_GROUP=-dev
-MEDIA_PATH=/data/sighthound/media
```
Comment (paroque28) [inline amqp-stats/conf/default.env:5]: Removed because now, there is a global variable  SIO_MEDIA_PATH for each service instead of copying and pasting the same content

File: examples/AqueductRunner/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,13 @@
+FROM python:3.8
```
Comment (paroque28) [inline examples/AqueductRunner/Dockerfile:1]: Aqueduct was added as an example and I've also included this example python script to use it with CLI

File: examples/AqueductRunner/src/aqueductRunner.py [hunk 3]
```diff
@@ -0,0 +1,439 @@
+#! /usr/bin/env python3
```
Comment (paroque28) [inline examples/AqueductRunner/src/aqueductRunner.py:1]: already reviewed in previous PR

File: scripts/sh-services [hunk 4]
```diff
@@ -1,11 +1,24 @@
 #!/bin/bash
-SCRIPT_PATH="$( cd -- "$(dirname "$0")" >/dev/null 2>&1 ; pwd -P )"
+SCRIPT_PATH="$(
+    cd -- "$(dirname "$0")" >/dev/null 2>&1
+    pwd -P
+)"
+export SCRIPT_PATH
```
Comment (paroque28) [inline scripts/sh-services:6]: These paths are also included in every .env file of each service

File: scripts/sh-services [hunk 5]
```diff
@@ -129,59 +166,222 @@ function merge(){
...
         fi
     done
-            
+
     merge_env_files "${service}"/.env ${env_files_to_merge}
 
 }
 
-function edit(){
+function set_text_editor() {
```
Comment (paroque28) [inline scripts/sh-services:180]: Select between vim and nano

File: README.md [hunk 6]
```diff
@@ -1,25 +1,53 @@
...
+This repository hosts a collection of services for the SIO (Sighthound Input/Output) ecosystem. Services are intended to run via the `./scripts/sh-services` script, which relies on docker-compose. However, docker-compose is not strictly necessary for configuring or using this repository.
+
+The included services are as follow:
+- SIO: The computer vision engine.
+- MCP: Media manager service, which includes a REST API and a cleaner.
+- RabbitMQ: Our AMQP broker and messaging platform.
+- Live555: An RTSP server for testing purposes. Note: This is disabled by default.
+- AMQP Stats:Connects to RabbitMQ and displays SIO output data. Note: This is disabled by default.
+
+The `./scripts/sh-services` script is a basic tool that triggers `docker-compose up/down` commands as needed. However, the primary logic lies in configuration management. This Command-Line Interface (CLI) tool reads the `conf` folder for each service and merges the configurations, based on alphanumeric priority, into the `.env` file used by docker-compose.
```
Comment (w3sip) [inline README.md:12]: Prepend with a paragraph:

In the turnkey scenario, each service is managed with an individual `docker-compose` configuration file, an optional `.env` file containing relevant environment variable, and a collection of service specific configuration file in `conf` subfolder. To assist in orchestrating the services collection and disjointed `docker-compose` and environment configuration, `sh-services` CLI utility was introduced.

File: README.md [hunk 7]
```diff
@@ -1,25 +1,53 @@
...
 
 ## Quick start
 
-This repo contains a set of docker-compose services that are meant to be handled using the `./scripts/sh-services` script. It is a simple tool that calls `docker-compose up/down` but the main logic resides in the configuration management. The CLI tool reads the `conf` folder of every service and performs a merge (using alphanumeric priority) into the `.env` file that docker-compose uses.
+This guide will help you set up SIO to point to a fake RTSP generated by live555 and start processing video.
 
+### Prerequisites (for non-dnncam devices)
 
-### Node devices
+If you are on a device that doesn't have `sh-device-ui` installed (dnn-cameras have this helper UI that sets the configuration) , you need to manually:
```
Comment (w3sip) [inline README.md:34]: Maybe rephrase:

On Sighthound DNNCam devices, `services` come preinstalled, and interact with the device GUI. We suggest you rely on the GUI to configure/update services, though it's not a requirement. On other devices, you need to manually:

File: README.md [hunk 8]
```diff
@@ -32,33 +60,33 @@ RELEASE="v1.3.0"
...
 
-#### Logging in to the docker registry
+#### Prerequisit 3: Logging into the Docker Registry
 
 ```bash
-docker login -u _json_key -p  "`cat /data/sighthound/keys/sighthound-keyfile.json`" us-central1-docker.pkg.dev
+docker login -u _json_key -p  "`cat "${SH_BASE}"/keys/sighthound-keyfile.json`" us-central1-docker.pkg.dev
 ```
 
 ### Enabling a test RTSP
```
Comment (w3sip) [inline README.md:92]: add `(in case you don't have other RTSP source available)`

File: sio/README.md [hunk 9]
```diff
@@ -1,5 +1,5 @@
 # SIO service Manual
-
+short description: SIO is the Sighthound Computer Vision Engine
 ## Limitations
 
 - NVMPI support is broken with nvidia-generated streams, and disabled for those. It is tested functional with most other H264 source we've tried, but if you experience problems, add `SIO_ENABLE_NVMPI=0` to your environment file.
```
Comment (w3sip) [inline sio/README.md:5]: This bit about NVMPI can be removed now

File: README.md [hunk 10]
```diff
@@ -1,25 +1,53 @@
-# Sighthound docker-compose service
+# Sighthound Services
+
+This repository hosts a collection of services for the SIO (Sighthound Input/Output) ecosystem. Services are intended to run via the `./scripts/sh-services` script, which relies on docker-compose. However, docker-compose is not strictly necessary for configuring or using this repository.
+
+The included services are as follow:
+- SIO: The computer vision engine.
+- MCP: Media manager service, which includes a REST API and a cleaner.
```
Comment (w3sip) [inline README.md:7]: Add:

MCP relies on sharing media store folders with SIO service, and listens on AMQP message bus for media creation events, such as new video recording segments or event-driven jpeg images. It then provides the API access to that media (for documentation please see [url]), and control its lifecycle.

**Review by w3sip — CHANGES_REQUESTED** [review]
Some initial comments. @orlandobrent - once we're done with the initial draft of cloud PR, would appreciate your review as well.

**Review by w3sip — CHANGES_REQUESTED** [review]
Some initial comments. @orlandobrent - once we're done with the initial draft of cloud PR, would appreciate your review as well.
